SPHINXsys/src/shared/particle_dynamics/solid_dynamics/contact_dynamics.cpp

```diff
@@ -77,52 +77,35 @@ namespace SPH
-			, particle_spacing_(solid_body_contact_relation.sph_body_.sph_adaptation_->ReferenceSpacing())
-			, calibration_factor_(StdVec<Real>(contact_configuration_.size(), 0.0))
-			, contact_h_ratio_(StdVec<Real>(contact_configuration_.size(), 0.0))
-			, offset_W_ij_(StdVec<Real>(contact_configuration_.size(), 0.0))
-			, contact_particle_spacing_(StdVec<Real>(contact_configuration_.size(), 0.0))
+			: LocalDynamics(solid_body_contact_relation.sph_body_),
+			  ContactDynamicsData(solid_body_contact_relation), solid_(particles_->solid_),
+			  kernel_(solid_body_contact_relation.sph_body_.sph_adaptation_->getKernel()),
+			  particle_spacing_(solid_body_contact_relation.sph_body_.sph_adaptation_->ReferenceSpacing())
 		{
 			particles_->registerVariable(contact_density_, "ContactDensity");
 
-			Real dp_1 = solid_body_contact_relation.sph_body_.sph_adaptation_->ReferenceSpacing();
 			for (size_t k = 0; k != contact_particles_.size(); ++k)
 			{
-				Real dp_2 = solid_body_contact_relation.contact_bodies_[k]->sph_adaptation_->ReferenceSpacing();
-				contact_particle_spacing_[k] = 0.5 * dp_1 + 0.5 * dp_2;
-				contact_h_ratio_[k] = dp_1 / contact_particle_spacing_[k];
-				offset_W_ij_[k] = kernel_->W(contact_h_ratio_[k], contact_particle_spacing_[k], zero_vec);
-				
-				contact_Vol_.push_back(&(contact_particles_[k]->Vol_));
-			}
+				Real dp_k = solid_body_contact_relation.contact_bodies_[k]->sph_adaptation_->ReferenceSpacing();
+				Real average_spacing_k = 0.5 * particle_spacing_ + 0.5 * dp_k;
+				Real h_ratio_k = particle_spacing_ / average_spacing_k;
+				offset_W_ij_.push_back(kernel_->W(h_ratio_k, average_spacing_k, ZeroVecd));
 
-			Real contact_max_;
-			for (size_t k = 0; k < contact_configuration_.size(); ++k)
-			{
-				for (int i = 0; i != 3; ++i)
+				Real contact_max(0.0);
+				Real contact_temp(0.0);
+				for (int l = 0; l != 3; ++l)
 				{
-					if (Dimensions == 2)
-					{
-						contact_max_ = 2.0 *
-							(kernel_->W(contact_h_ratio_[k], three_gaussian_points_[i] * contact_particle_spacing_[k] * 0.5 + contact_particle_spacing_[k] * 0.5, Vec2d{0,0})
-								- offset_W_ij_[k])
-							* contact_particle_spacing_[k] * 0.5 * three_gaussian_weights_[i];
-					}
-					else
-					{
-						contact_max_ =
-							(kernel_->W(contact_h_ratio_[k], three_gaussian_points_[i] * contact_particle_spacing_[k] * 0.5 + contact_particle_spacing_[k] * 0.5, Vec3d{0,0,0})
-								- offset_W_ij_[k])
-							* 2.0 * Pi * (three_gaussian_points_[i] * contact_particle_spacing_[k] * 0.5 + contact_particle_spacing_[k] * 0.5)
-							* contact_particle_spacing_[k] * 0.5 * three_gaussian_weights_[i];
-					}
+					Real temp = three_gaussian_points_[l] * average_spacing_k * 0.5 + average_spacing_k * 0.5;
+					contact_temp = 2.0 * (kernel_->W(h_ratio_k, temp, ZeroVecd) - offset_W_ij_[l]) *
+						average_spacing_k * 0.5 * three_gaussian_weights_[l];
+					contact_max += contact_temp;
+					if (Dimensions == 3)
```

Comment:
should it be contact_max += Dimensions == 3 ? contact_temp : 	contact_temp * Pi * temp ?

Reply:
> should it be contact_max += Dimensions == 3 ? contact_temp : contact_temp * Pi * temp ?

Yes, better. Thank you. Seems like the regression test for 3d_elasticSolid_shell_collision has the problem. I will correct that together.